worker/credentialvalidator/util_test.go

```diff
@@ -0,0 +1,153 @@
+func (m *mockFacade) ModelCredential() (base.StoredCredential, bool, error) {
+	m.AddCall("ModelCredential")
+	if err := m.NextErr(); err != nil {
+		return base.StoredCredential{}, false, err
+	}
+	return m.nextCredential(), m.exists, nil
+}
+
+// nextCredential consumes a credential and returns it, or panics.
+func (m *mockFacade) nextCredential() base.StoredCredential {
+	credential := m.credentials[0]
+	m.credentials = m.credentials[1:]
+	return credential
+}
+
+// WatchCredential is part of the credentialvalidator.Facade interface.
+func (mock *mockFacade) WatchCredential(id string) (watcher.NotifyWatcher, error) {
+	mock.AddCall("WatchCredential", id)
+	if err := mock.NextErr(); err != nil {
+		return nil, err
+	}
+	return newMockWatcher(), nil
+}
+
+// newMockWatcher returns a watcher.NotifyWatcher that always
+// sends 3 changes and then sits quietly until killed.
+func newMockWatcher() *mockWatcher {
+	const count = 3
+	changes := make(chan struct{}, count)
+	for i := 0; i < count; i++ {
+		changes <- struct{}{}
+	}
+	return &mockWatcher{
+		Worker:  workertest.NewErrorWorker(nil),
+		changes: changes,
+	}
+}
+
+// mockWatcher implements watcher.NotifyWatcher for use in the tests.
+type mockWatcher struct {
+	worker.Worker
+	changes chan struct{}
+}
+
+// Changes is part of the watcher.NotifyWatcher interface.
+func (mock *mockWatcher) Changes() watcher.NotifyChannel {
+	return mock.changes
+}
+
+// credentialTag is the credential tag we're using in the tests.
+// needs to fit fmt.Sprintf("%s/%s/%s", cloudName, userName, credentialName)
+var credentialTag = names.NewCloudCredentialTag("cloud/user/credential").String()
+
+// panicCheck is a Config.Check value that should not be called.
+func panicCheck(base.StoredCredential) bool { panic("unexpected") }
+
+// neverCheck is a Config.Check value that always returns false.
+func neverCheck(base.StoredCredential) bool { return false }
+
+// panicFacade is a NewFacade that should not be called.
+func panicFacade(base.APICaller) (credentialvalidator.Facade, error) {
+	panic("panicFacade")
+}
+
+// panicWorker is a NewWorker that should not be called.
+func panicWorker(credentialvalidator.Config) (worker.Worker, error) {
+	panic("panicWorker")
+}
+
+// validConfig returns a minimal config stuffed with dummy objects that
+// will explode when used.
+func validConfig() credentialvalidator.Config {
+	return credentialvalidator.Config{
+		Facade: struct{ credentialvalidator.Facade }{},
+		Check:  panicCheck,
+	}
+}
+
+// checkNotValid checks that the supplied credentialvalidator.Config fails to
+// Validate, and cannot be used to construct a credentialvalidator.Worker.
+func checkNotValid(c *gc.C, config credentialvalidator.Config, expect string) {
+	check := func(err error) {
+		c.Check(err, gc.ErrorMatches, expect)
+		c.Check(err, jc.Satisfies, errors.IsNotValid)
+	}
+
+	err := config.Validate()
+	check(err)
+
+	worker, err := credentialvalidator.New(config)
```

Comment:
You only need to check that the config is validated in the new worker function once. This violates the "test it once" idea.